### DISCUSSION TITLE
Unable to get the transport ssl context from the request. This prevents checking the Client provided certificate and matching up the provided CN against allowed users/server.

### DISCUSSION THREAD

**Discussion by desean1625** [Kludex/uvicorn#2307] [Q&A]
### Initial Checks

- [X] I confirm this was discussed, and the maintainers suggest I open an issue.
- [X] I'm aware that if I created this issue without a discussion, it may be closed without a response.

### Discussion Link

```Text
https://github.com/encode/uvicorn/issues/745
```

### Description

Many applications in finance/banking require two way certificate verification. Currently the way we have handled this is by proxying the request and extracting out the client information at nginx or traefik and stuffing it into the headers.

### Example Code

From the request we cannot get the transport information and unable to getgetpeercert preventing application-level validation of client certificates.

A possible solution is to pass the transport in the request scope.
In the protocol h11_impl.py we could simply add
```py
"transport":self.transport
```
after
https://github.com/encode/uvicorn/blob/0efd3835da6dcc713f74aadf7b52779d0d1fa17d/uvicorn/protocols/http/h11_impl.py#L203

Then at a route level or fastapi middleware we could pull the client certificates to check against an authorization service.

```py
@app.get("/admin")
async def getAdminPage(request:Request):
  client_cert = request.scope['transport'].get_extra_info("ssl_object").getpeercert()
  #Verify user common name is an admin

```

### Python, Uvicorn & OS Version

```Text
All
```

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/uvicorn/issues/2306">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/uvicorn/issues/2306/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/uvicorn/issues/2306/pledge.svg">
</picture>
</a>

**Comment by lschneidpro** [comment]
any update ?

**Comment by frobnitzem** [comment]
I agree, and suggest another issue name:

## Pass underlying transport via Scope["extensions"]

In order for an API function to look into properties of the socket/transport that called it, the preferred method is via scope extensions.
This is needed, for example, to get the client TLS certificate using `transport.get_extra_info('peercert')`  [ref](https://github.com/tiangolo/fastapi/issues/2224).

The basic ASGI call sends scope, along with receive and send callbacks:
https://github.com/encode/uvicorn/blob/5bf788f0eb0fc771f5c4eed8f282c7ec256565d2/uvicorn/protocols/http/httptools_impl.py#L397-L401

Scope, defined in the [ASGI spec](https://asgi.readthedocs.io/en/latest/specs/main.html), allows "extensions" to be defined by implementations.

So, I also agree we should set `scope["extensions"]["transport"] = self.transport`
here: https://github.com/encode/uvicorn/blob/5bf788f0eb0fc771f5c4eed8f282c7ec256565d2/uvicorn/protocols/http/httptools_impl.py#L213-L223

**Reply by frobnitzem** [reply]
Note - a "tls" extension has been implemented and is in need of a final review here:
https://github.com/encode/uvicorn/pull/1119. nonecorn and hypercorn are also mentioned in that thread.

Sending the transport through is still useful, since there is still more information on the socket -- for example with `AF_UNIX` sockets, this allows reading the username:
`peer = transport.get_extra_info('socket').getsockopt(socket.SOL_SOCKET, socket.SO_PEERCRED, struct.calcsize("3I"))`

**Reply by desean1625** [reply]
This is great. It looks like it should work.

Currently I have just been monkey patching the RequestResponseCycle
```py
from uvicorn.protocols.http.h11_impl import RequestResponseCycle
responseCycleInit = RequestResponseCycle.__init__
def monkey_patch_response_cycle(self,*k,**kw):
  responseCycleInit(self,*k,**kw)
  self.scope['transport'] = self.transport

RequestResponseCycle.__init__ = monkey_patch_response_cycle

```

**Comment by schlitzered** [comment]
Hi,

i also would like to see the [ASGI TLS Extension](https://asgi.readthedocs.io/en/latest/specs/tls.html) implemented in uvicorn. i need this for mtls to be able to inspect the client cert that was provided.

I prepared this [PR](https://github.com/Kludex/uvicorn/pull/3043) that was closed, i guess because i did not start a discussion first, sorry for that.


@Kludex is this something you think is a meaningful feature for uvicorn? i would like to get rid of me current subclasses for the http11 and websocket protocol that add this info.
